### PR TITLE
Make sure that parents appear before their children when constructing symbol tree

### LINKED ISSUE #4060: Types declared after their members not shown correctly in symbol tree
See https://github.com/geany/geany/pull/4039#issuecomment-2486990414 and the following posts.

### PR DESCRIPTION
**Pull request by techee**
The symbol tree construction code assumes that tag parents appear before their children. At the moment tag list is sorted by line numbers which works for C structs
```C
struct Foo {
    int bar;
    int baz;
}
```
where the tag Foo (which is a parent of bar and baz) is sorted before bar and baz. However, this doesn't work when the name of the struct (or a variable of an anonymous struct) follows the struct members such as in the following SystemVerilog example:
```systemverilog
struct packed {
    byte s_a, s_b, s_c;
    string s_str;
} s_member;
```
Here the members are sorted before the parent and the symbol tree construction code doesn't work correctly.

This patch sorts symbols so that parents are sorted before their members - only if tags aren't in a parent-child relationship the original sorting by line is performed.

@b4n Does this look alright to you?

Fixes #4060.

### CODE REVIEW

**Comment by cousteaulecommandant** [comment]
This seems to work with the example file I created (`tests/ctags/oop.sv`), and also with a small example I made.  Thanks!

I found that this doesn't work if I declare the struct or typedef struct outside of a module, package, etc; but I'm not even sure that's legal in SystemVerilog.  Probably not, so I wouldn't worry.

**Comment by techee** [comment]
> I found that this doesn't work if I declare the struct or typedef struct outside of a module, package, etc;

Could you give a short example of such code?

**Comment by techee** [comment]
> Could you give a short example of such code?

Never mind, I can see it too. I'll look into this.

**Comment by techee** [comment]
Should be fixed now, thanks for noticing.

**Comment by techee** [comment]
Just a note about the implementation - I wanted to avoid allocation/deallocation in the compare function because of possible performance issues otherwise the code would use concatenation and be a bit simpler.

**Comment by cousteaulecommandant** [comment]
Yeah that fixed it :) Thanks!

File: src/symbols.c [hunk 1]
```diff
@@ -324,6 +324,56 @@ static gint compare_symbol_lines(gconstpointer a, gconstpointer b)
...
+		if (!g_str_has_prefix(tag->scope + scope_len, scope_sep))
+			return FALSE;
+		scope_len += strlen(scope_sep);
+	}
+
+	if (!g_str_has_prefix(tag->scope + scope_len, parent->name))
+		return FALSE;
+	scope_len += strlen(parent->name);
+
+	if (scope_len == strlen(tag->scope))
```
Comment (b4n) [inline src/symbols.c:350]: Nitpick: I'd rather write `! tag->scope[scope_len]` or one of the many equivalents that avoid yet another `strlen()`

**Review by b4n — COMMENTED** [review]
I have yet to test this, but it looks good and might fix those corner cases for several lesser tested languages 👍 

Just note that IIUC ctags tries to avoid emitting tags in the "wrong" order to avoid these kind of issues, so it might be worth mentioning there, although I don't known if it's actually worth fixing there given others can do something similar to what you suggest here.

File: src/symbols.c (hunk 1 above)
Comment (techee) [reply src/symbols.c:350]: Yes, this one crossed my mind before falling asleep yesterday :-)

**Comment by techee** [comment]
@b4n OK to merge this one?

**Comment by elextr** [comment]
I don't have a time to try it, but what I consider is better try Rust and C++ where names order and name visibility and lots of other.

**Review by b4n — CHANGES_REQUESTED** [review]
Apart from the small nitpick from my compiler, looks good and seems to indeed work nicely (tested that it e.g. fixed *tests/ctags/oop.sv*, and couldn't find any regression) :+1:

**Comment by techee** [comment]
> Apart from the small nitpick from my compiler, looks good and seems to indeed work nicely (tested that it e.g. fixed tests/ctags/oop.sv, and couldn't find any regression) 👍

Thanks, I added your suggestion and committed the squashed version.
